### ISSUE TITLE
Proposal for switching i18n method over to .PO/.POT/.MO files (GNU Gettext)

### ISSUE THREAD

**Issue by sexybiggetje** [ifmeorg/ifme#719]
TL:DR; GNU Gettext is an industry standard for i18n and has been developed for many years, many translating agencies/independent translators can work with them. Also switching to Gettext allows for easier updating of all translations, as software can update missing keys and perhaps even fill in blanks.

--

When using the Gettext approach one can define a master template, this is your .POT file. It is usually automatically generated from source code by corresponding tools.  For ruby there is the gettext gem, it contains tools and library for working with these files.

Translations are generated by command line tools, or by an application such as POEdit. They all run on many platforms, and the latter can provide some deduplication and automatic translation using online resources. When using Gettext one can also use online translation tools like Transifex (Free for FOSS), which opens up the world of translation to basically anyone.

Now why is this important? At the moment there are not that many languages, but adding/modifying strings and/or keys will become much of a hassle when >10 languages. Just imagine a dev needing to add the missing keys to all languages or having translators add them.

I'm no Ruby expert, but I have a long standing relationship with translating open source software and using POEdit.

**Comment by sexybiggetje** [comment]
I'm going to work on this in the following steps:
- Find out where translations are being done now (for instance t() calls in *.erb)
- Try to make a gettext extractor to generate the base .pot files
- Try to make a generator that matches .pot msgid's and generate .po files for all languages, so we don't need to copy paste

Then when we have these resources, we should look at how to load them and replace the current localisation. The .po/.pot format is somewhat open, but largely specified here:
http://pology.nedohodnik.net/doc/user/en_US/ch-poformat.html

I might opt-in to do the work for these extractors/generators in my [ifme-languagetools](https://github.com/sexybiggetje/ifme-languagetools) repository.

**Comment by timofonic** [comment]
@sexybiggetje I agree. This might make it easier to use web tools to translate it, so less computer literate people can contribute to translations too.

There are standards like XLIFF too, this one based on XML. It's possible to convert between XLIFF and PO, but not sure if some kind of data is lost in the process.


Incomplete example list..
- Web apps:
  * [Weblate](https://weblate.org): Developed in Python using Django framework. It can be easily integrated in places like GitHub and GitLab. It's [Free Software](https://github.com/WeblateOrg/weblate) and can be self-hosted.
  * [Pootle](https://github.com/translate/): It's quite popular, but less than Weblate these days (I don't know why).
  * [Zanata](http://zanata.org): A [Free Software](https://github.com/zanata) solution by Red Hat. Developed in Java.
  * [Translatewiki](https://translatewiki.net): It's based on different MediaWiki extensions.

- Software:
  * [OmegaT](http://www.omegat.org). It's [Free Software](https://github.com/omegat-org/). It supports lots of formats and can be used along other tools (like Oktopi ones). Developed in Java.
  * [Xliffie](https://hotmilktea.com/xliffie/): A [Free Software](https://github.com/mrahmiao/xliffie) XLIFF editor.  It got abandoned, but recently forked. Developed in Objective-C.
  * [https://www.kde.org/applications/development/lokalize/](Lokalize): Developed by KDE project. It supports PO, XLIFF.
* [Okapi Framework](http://okapiframework.org): It's a [Free Software](https://bitbucket.org/okapiframework/) framework. There are lots of tools that use it like [Rainbow](http://okapiframework.org/wiki/index.php?title=Rainbow),  [CheckMate](http://okapiframework.org/wiki/index.php?title=CheckMate), [Ratel](http://okapiframework.org/wiki/index.php?title=Ratel), [Tikal](http://okapiframework.org/wiki/index.php?title=Tikal),  Filters Plugin for OmegaT , [Longhorn](http://okapiframework.org/wiki/index.php?title=Longhorn), [Ocelot](https://github.com/vistatec/ocelot), Okapi Components for WorldServer. Developed in Java.
  * [Virtaal](https://github.com/translate/virtaal): It's somewhat abandoned. Developed in Python.

For videos there's [Amara](https://amara.org), that is [Free Software (see repo)](https://github.com/pculture/unisubs).

There's standards like [Translation Web Services](https://www.oasis-open.org/committees/tc_home.php?wg_abbrev=trans-ws) that might make everything easier too.

**Comment by camillevilla** [comment]
Sorry! Not sure why that dependabot PR closed this. 😅
